src/librustc/infer/const_variable.rs

```diff
@@ -0,0 +1,271 @@
+use crate::mir::interpret::ConstValue;
+use syntax::symbol::InternedString;
+use syntax_pos::Span;
+use crate::ty::{self, InferConst};
+
+use std::cmp;
+use std::marker::PhantomData;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::snapshot_vec as sv;
+use rustc_data_structures::unify as ut;
+
+pub struct ConstVariableTable<'tcx> {
```

Comment:
We had something more basic initially, but kept running into situations where an analogous abstraction to `TypeVariableTable` didn't exist. Replicating a similar data structure for consts made everything clearer, but it's possible that some of what we were trying to do is simply unnecessary at that level of generality for consts.

> The problem with types is that there are "integer infer variables" and "float infer variables", but constants only have one kind of inference variable.

Is this not a separate concern (i.e. there's `type_variables`, `int_unification_table` and `float_unification_table` and at the moment, `const_variables` follows the structure laid out by `type_variables` rather than the other two)? It might be easiest to identify places where the const unification table is doing more work than necessary, so we can work out if it isn't actually necessary.

Reply:
Oh I just thought there is already something like this, then I don't know.

Is there no `..._table` field that has a type that's... not from librustc? Some general-purposes "unification table"?

Reply:
I don't see one in `InferCtxt` at least. It could make sense to create an abstraction if it doesn't exist (and I'm not just overlooking it somewhere) as a `C-cleanup` issue afterwards.

Reply:
I just checked and it turns out that I was referring to `int_unification_table` and `float_unification_table`, which rely on e.g.:
https://github.com/rust-lang/rust/blob/aa97448caf49ee25271a5ad58482283404b9144f/src/librustc/infer/unify_key.rs#L8-L13

The reason this wasn't immediately obvious was that you have a `values` field and that it *doesn't* hold the actual variable values (maybe it used to, for `TypeVariableTable`, at some point?) *but it's really* "origins" (and should be named as such; in `TypeVariableTable` too).

Note that the `UnificationTable` holds `<Vid as UnifyKey>::Value` for each active vid, and you could combine `ConstVariableData` and `ConstVariableValue` into one type to use as that `Value` associated type.

A bit surprised to see integer/float inference variables don't have an origin attached, but they could all use the same mechanism.

IMO the reason `type_variables` exists as a separate module is mainly because of subtyping relations, and we shouldn't need that here.